File: tests/_test_automodule.cc
#include "libpy/autoclass.h"
#include "libpy/autofunction.h"
#include "libpy/automodule.h"

bool is_42(int arg) {
    return arg == 42;
}

bool is_true(bool arg) {
    return arg;
}

using int_float_pair = std::pair<int, float>;

int first(const int_float_pair& ob) {
    return ob.first;
}

float second(const int_float_pair& ob) {
    return ob.second;
}

LIBPY_AUTOMODULE(tests,
                 _test_automodule,
                 ({py::autofunction<is_42>("is_42"),
                   py::autofunction<is_true>("is_true")}))
(py::borrowed_ref<> m) {
    py::autoclass<int_float_pair>(m, "int_float_pair")
        .new_<int, float>()
        .comparisons<int_float_pair>()
        .def<first>("first")
        .def<second>("second")
        .type();
    return false;
}

File: tests/test_automodule.py
from . import _test_automodule as mod


def test_modname():
    assert mod.__name__ == 'tests._test_automodule'


def test_function():
    assert mod.is_42(42)
    assert not mod.is_42(~42)


def test_type():
    assert isinstance(mod.int_float_pair, type)
    a = mod.int_float_pair(1, 2.5)
    assert a.first() == 1
    assert a.second() == 2.5
    b = mod.int_float_pair(1, 2.5)
    assert a == b
    c = mod.int_float_pair(1, 3.5)
    assert a != c
